include/arch/arm/cortex_m/asm_inline_gcc.h

```diff
@@ -169,9 +169,16 @@ static ALWAYS_INLINE void _arch_irq_unlock(unsigned int key)
 	if (key) {
 		return;
 	}
-	__asm__ volatile("cpsie i" : : : "memory");
+	__asm__ volatile(
+		"cpsie i;"
+		"isb"
+		: : : "memory");
 #elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
-	__asm__ volatile("msr BASEPRI, %0" :  : "r"(key) : "memory");
+	__asm__ volatile(
+		"msr BASEPRI, %0;"
+		"dsb;"
```

Comment:
Do we need DSB here? The BASEPRI is a CPU register accessed through special instructions, so it is not affected by the memory ordering rules.

Reply:
Please, refer to the 321 application note (link attached in PR), in the special case of Enabling interrupts with CPSIE/MSR.
True, BASEPRI is in Strongly-ordered memory area (SCB) but this does not ensure relative ordering with normal-memory (e.g. could be grouped with), so if we really need the effect of interrupts' enable to be realized before any more instructions are executed, we need barriers. The Note is clear about that, at least.

Reply:
From the mentioned document:
> CPS and MSR instructions
> In addition to barriers, the CPSID instruction is also self synchronizing within the instruction stream, and ensures that the effect of setting PRIMASK and/or FAULTMASK is visible on the following instruction, thus ensuring that the appropriate interrupt/exceptions are disabled

So data-related barriers are needed after STR, but not after MSR.

Reply:
This PR is about interrupt-enable, not disable.

Reply:
Check CPSIE section.

Reply:
@ioannisg: I have updated my post. please re-read.

Reply:
>So data-related barriers are needed after STR, but not after MSR, which omits all ordering logic.

Correct. But this is about DMB. I have not added DMBs here, anyways.

Reply:
>>The same requirement applies when using the MSR instruction to enable interrupts.

So, thisis tricky. It refers to MSR PRIMASK and/or FAULTMASK, but BASEPRI is _strictly_ not a register that enables or disabled interrupts. It sets the base level only. So strictly, the text you added does not cover this!

As a result, I added DSB to be completely sure it aligns with other references in this Note about "recognizing" SCB writes.

Reply:
But DSB is not needed as you are writing using MSR not STR :)

Reply:
So, here is what I see from the specification (Generic case):
>>The DSB instruction ensures all explicit data transfers before the DSB are complete before any instruction after the DSB is executed. In Cortex-M processors, it can be used to:
>>- ensure effects of an access to SCS take place before the next operation
>>- ensure memory is updated before the next operation, for example, SVC, WFI, WFE.In an application using CMSIS, you can use _DSB() function to insert the DSB instruction

BASEPRI _is_ implemented in SCB, so, in general, DSB is needed for synchronization.

Reply:
OK, I realize now what you mean.

Reply:
Yes, you are right (I think). The BASEPRI is, indeed, implemented in SCB, but programmed with MSR - which should include DSB nature. I will remove this. Thanks

Reply:
No. The BASEPRI is a core register, like the CONTROL, R0-R15, PSP etc. This is why you have to use MRS/MSR to access it.
If it would be system control register, it would have address and you would use the LDR/STR instructions to access because system controll is mapped to Private Peripheral Bus (0xE0000000- 0xE00FFFFF). In such case the barriers would be needed.

Please look at the http://infocenter.arm.com/help/topic/com.arm.doc.dui0553b/DUI0553.pdf, section 2.1.3. Note that the registers does not have addresses. Then please look at section 4.3. Please note that addresses are listed there.

Reply:
Right. I got really confused about this. It's really clear.